sql/core/src/main/scala/org/apache/spark/sql/execution/joins/OuterJoin.scala

```diff
@@ -111,52 +122,43 @@ trait HashOuterJoin {
   // iterator for performance purpose.
 
   protected[this] def leftOuterIterator(
-      key: InternalRow,
       joinedRow: JoinedRow,
       rightIter: Iterable[InternalRow],
       resultProjection: InternalRow => InternalRow): Iterator[InternalRow] = {
     val ret: Iterable[InternalRow] = {
-      if (!key.anyNull) {
-        val temp = if (rightIter != null) {
-          rightIter.collect {
-            case r if boundCondition(joinedRow.withRight(r)) => resultProjection(joinedRow).copy()
```

Comment:
Yeah, just noticed the same thing :(

This is one reason why I like some of the new internal iterator-like interfaces that we have, which fuse together the `hasNext()` and `getNext()` calls into a single `next()` method that returns a Boolean. Short of using our own iterator implementations throughout Spark SQL I think that we're going to be stuck performing this copy here.

However, we still might be able to improve efficiency by avoiding the `collect()` and the use of partial functions here.